kratos/containers/distributed_sparse_graph.h

```diff
@@ -0,0 +1,427 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+#if !defined(KRATOS_DISTRIBUTED_SPARSE_GRAPH_H_INCLUDED )
+#define  KRATOS_DISTRIBUTED_SPARSE_GRAPH_H_INCLUDED
+
+
+// System includes
+#include <iostream>
+#include "includes/ublas_interface.h"
+#include "includes/serializer.h"
+#include "includes/parallel_environment.h"
+#include "utilities/communication_coloring_utilities.h"
+
+// External includes
+#include <unordered_map>
+#include <unordered_set>
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+//class to construct and store a matrix graph. Can be used to construct efficiently a CSR matrix (or other sparse matrix types)
+
+/** This class is designed to store a matrix graph, aimed at the fast construction of other
+ * sparse matrix formats (particularly CSR)
+ * IMPORTANT NOTE: it is BY DESIGN NOT threadsafe! (a graph should be computed in each thread and then merged)
+*/
+
+class DistributedSparseGraph //: public SparseGraph
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    typedef std::size_t IndexType; //note that this could be different from the one in the basetype
```

Comment:
The thing is that i believe that we should use 32 bit integers for the sparsegraph and 64bits integers for the distributed one. of course to be discussed

Reply:
in openmp mode i think we will never exceed 2,147,483,647 entries in a sparse matrix (the double if we use unsigned int).

**i agree about using the current default,** however please note that this change would reduce drastically (25%) the occupation of sparse matrices and of preconditioners.
Also using a shorter index type would probably allow some performance gain because of better cache usage.
In short, i would like to try this out at a later stage